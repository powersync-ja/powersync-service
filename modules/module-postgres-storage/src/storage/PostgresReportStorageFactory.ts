import { storage } from '@powersync/service-core';
import * as pg_wire from '@powersync/service-jpgwire';
import { event_types } from '@powersync/service-types';
import { v4 } from 'uuid';
import * as lib_postgres from '@powersync/lib-service-postgres';
import { NormalizedPostgresStorageConfig } from '../types/types.js';

import { STORAGE_SCHEMA_NAME } from '../utils/db.js';
import { getStorageApplicationName } from '../utils/application-name.js';
import {
  DeleteOldSdkData,
  ListCurrentConnections,
  ListCurrentConnectionsRequest,
  ScrapeSdkDataRequest,
  SdkConnectBucketData,
  SdkDisconnectEventData
} from '@powersync/service-types/src/events.js';

export type PostgresReportStorageOptions = {
  config: NormalizedPostgresStorageConfig;
};

export class PostgresReportStorageFactory implements storage.ReportStorageFactory {
  readonly db: lib_postgres.DatabaseClient;
  constructor(protected options: PostgresReportStorageOptions) {
    this.db = new lib_postgres.DatabaseClient({
      config: options.config,
      schema: STORAGE_SCHEMA_NAME,
      applicationName: getStorageApplicationName()
    });

    this.db.registerListener({
      connectionCreated: async (connection) => this.prepareStatements(connection)
    });
  }

  private parseJsDate(date: Date) {
    const year = date.getFullYear();
    const month = date.getMonth();
    const today = date.getDate();
    const day = date.getDay();
    return {
      year,
      month,
      today,
      day,
      parsedDate: date
    };
  }
  private timeFrameQuery(timeframe: event_types.TimeFrames, interval: number = 1) {
    const { year, month, today, parsedDate } = this.parseJsDate(new Date());
    const parsedIsoString = parsedDate.toISOString();
    switch (timeframe) {
      case 'month': {
        return { lt: parsedIsoString, gt: new Date(year, parsedDate.getMonth() - interval).toISOString() };
      }
      case 'week': {
        const weekStartDate = new Date(parsedDate);
        weekStartDate.setDate(weekStartDate.getDate() - 6 * interval);
        const weekStart = this.parseJsDate(weekStartDate);
        return {
          lt: parsedIsoString,
          gt: new Date(weekStart.year, weekStart.month, weekStart.today).toISOString()
        };
      }
      case 'hour': {
        // Get the last hour from the current time
        const previousHour = parsedDate.getHours() - interval;
        return {
          lt: new Date(year, month, today, parsedDate.getHours()).toISOString(),
          gt: new Date(year, month, today, previousHour).toISOString()
        };
      }
      default: {
        return {
          lt: parsedIsoString,
          gt: new Date(year, month, today - interval).toISOString()
        };
      }
    }
  }

  private timeFrameDeleteQuery(timeframe: event_types.TimeFrames, interval: number = 1) {
    const { year, month, today, parsedDate } = this.parseJsDate(new Date());
    switch (timeframe) {
      case 'month': {
        return { lt: new Date(year, parsedDate.getMonth() - interval).toISOString() };
      }
      case 'week': {
        const weekStartDate = new Date(parsedDate);
        weekStartDate.setDate(weekStartDate.getDate() - 6 * interval);
        const { month, year, today } = this.parseJsDate(weekStartDate);
        return {
          lt: new Date(year, month, today).toISOString()
        };
      }
      case 'hour': {
        const previousHour = parsedDate.getHours() - interval;
        return {
          lt: new Date(year, month, today, previousHour).toISOString()
        };
      }
      default: {
        return {
          lt: new Date(year, month, today - interval).toISOString()
        };
      }
    }
  }

  private listConnectionsDateRangeQuery(data: event_types.ListCurrentConnectionsRequest) {
    const { range } = data;
    if (!range) {
      const query = `
        WITH filtered AS (
              SELECT *
              FROM sdk_report_events
              WHERE disconnect_at IS NULL
              AND jwt_exp > NOW()
                        ),
        unique_users AS (
              SELECT COUNT(DISTINCT user_id) AS count
              FROM filtered
                        ),
        sdk_versions_array AS (
              SELECT sdk,
              COUNT(*) AS total,
              COUNT(DISTINCT client_id) AS clients,
              COUNT(DISTINCT user_id) AS users
              FROM filtered
              GROUP BY sdk
                            )
        SELECT
          COALESCE(u.count, 0) AS users,
          JSON_AGG(ROW_TO_JSON(s)) AS sdks
        FROM unique_users u
        JOIN sdk_versions_array s ON TRUE;
    `;
      return {
        statement: query
      };
    }
    const endDate = data.range?.end_date ? new Date(data.range.end_date) : new Date();
    const startDate = new Date(range.start_date);
    const query = `
      WITH filtered AS (
              SELECT *
              FROM sdk_report_events
              WHERE disconnect_at IS NULL
              AND jwt_exp > NOW()
              AND connect_at > CAST($1 AS TIMESTAMP WITH TIME ZONE)
              AND connect_at <= CAST($2 AS TIMESTAMP WITH TIME ZONE)
      ),
      unique_users AS (
              SELECT COUNT(DISTINCT user_id) AS count
              FROM filtered
                        ),
      sdk_versions_array AS (
              SELECT sdk,
              COUNT(*) AS total,
              COUNT(DISTINCT client_id) AS clients,
              COUNT(DISTINCT user_id) AS users
              FROM filtered
              GROUP BY sdk
                            )
      SELECT COALESCE(u.count, 0) AS users, JSON_AGG(ROW_TO_JSON(s)) AS sdks
      FROM unique_users u
      JOIN sdk_versions_array s ON TRUE;
    `;
    const lt = endDate.toISOString();
    const gt = startDate.toISOString();
    return { statement: query, params: [{ value: gt }, { value: lt }] };
  }

  private updateTableFilter() {
    const { year, month, today } = this.parseJsDate(new Date());
    const nextDay = today + 1;
    return {
      gte: new Date(year, month, today).toISOString(),
      lt: new Date(year, month, nextDay).toISOString()
    };
  }

  async reportSdkConnect(data: SdkConnectBucketData): Promise<void> {
    const { sdk, connect_at, user_id, user_agent, jwt_exp, client_id } = data;
    const connectIsoString = connect_at.toISOString();
    const jwtExpIsoString = jwt_exp!.toISOString();
    const { gte, lt } = this.updateTableFilter();
    const query = `
    INSERT INTO sdk_report_events (user_id, client_id, connect_at, sdk, user_agent, jwt_exp, id)
    VALUES ($1, $2, $3, $4, $5, $6, $7)
    ON CONFLICT (user_id, client_id, connect_at)
    DO UPDATE SET
      connect_at = CAST($3 AS TIMESTAMP WITH TIME ZONE),
      sdk = $4,
      user_agent = $5,
      jwt_exp = CAST($6 AS TIMESTAMP WITH TIME ZONE),
      disconnect_at = NULL
    WHERE sdk_report_events.connect_at >= CAST($8 AS TIMESTAMP WITH TIME ZONE)
      AND sdk_report_events.connect_at < CAST($9 AS TIMESTAMP WITH TIME ZONE);`;
    try {
      const result = await this.db.query({
        statement: query,
        params: [
          { type: 'varchar', value: user_id },
          { type: 'varchar', value: client_id },
          { type: 1184, value: connectIsoString },
          { type: 'varchar', value: sdk },
          { type: 'varchar', value: user_agent },
          { type: 1184, value: jwtExpIsoString },
          { type: 'varchar', value: v4() },
          { type: 1184, value: gte },
          { type: 1184, value: lt }
        ]
      });
      console.log(result.rows);
    } catch (error) {
      console.log(error);
    }
  }
  async reportSdkDisconnect(data: SdkDisconnectEventData): Promise<void> {
    const { user_id, client_id, disconnect_at } = data;
    const disconnectIsoString = disconnect_at.toISOString();
    const { gte, lt } = this.updateTableFilter();
    const query = `
      UPDATE sdk_report_events
      SET
        disconnect_at = CAST($1 AS TIMESTAMP WITH TIME ZONE),
        jwt_exp = NULL
      WHERE user_id = $2
        AND client_id = $3
        AND connect_at >= CAST($4 AS TIMESTAMP WITH TIME ZONE)
        AND connect_at < CAST($5 AS TIMESTAMP WITH TIME ZONE);`;

    try {
      const result = await this.db.query({
        statement: query,
        params: [
          { type: 1184, value: disconnectIsoString },
          { type: 'varchar', value: user_id },
          { type: 'varchar', value: client_id },
          { type: 1184, value: gte },
          { type: 1184, value: lt }
        ]
      });
      console.log(result.rows);
    } catch (error) {
      console.log(error);
    }
  }
  async listCurrentConnections(data: ListCurrentConnectionsRequest): Promise<ListCurrentConnections> {
    const statement = this.listConnectionsDateRangeQuery(data);
    const result = await this.db.query(statement);
    console.log(result.rows);
    return {
      users: 0,
      sdks: []
    };
  }

  async scrapeSdkData(data: ScrapeSdkDataRequest): Promise<ListCurrentConnections> {
    const { timeframe, interval } = data;
    const { lt, gt } = this.timeFrameQuery(timeframe, interval);
    const query = `
      WITH filtered AS (
            SELECT *
            FROM sdk_report_events
            WHERE connect_at > CAST($1 AS TIMESTAMP WITH TIME ZONE)
            AND connect_at <= CAST($2 AS TIMESTAMP WITH TIME ZONE)
                        ),
      unique_users AS (
            SELECT COUNT(DISTINCT user_id) AS count
            FROM filtered
                        ),
      sdk_versions_array AS (
            SELECT sdk,
            COUNT(*) AS total,
            COUNT(DISTINCT client_id) AS clients,
            COUNT(DISTINCT user_id) AS users
            FROM filtered
            GROUP BY sdk
                            )
      SELECT COALESCE(u.count, 0) AS users, JSON_AGG(ROW_TO_JSON(s)) AS sdks
      FROM unique_users u
      JOIN sdk_versions_array s ON TRUE;
    `;
    const result = await this.db.query({ statement: query, params: [{ value: gt }, { value: lt }] });
    console.log(result.rows);
    return {
      users: 0,
      sdks: []
    };
  }
  async deleteOldSdkData(data: DeleteOldSdkData): Promise<void> {
    const { timeframe, interval } = data;
    const { lt } = this.timeFrameDeleteQuery(timeframe, interval);
    const query = `
    DELETE FROM sdk_report_events
    WHERE connect_at < CAST($1 AS TIMESTAMP WITH TIME ZONE)
      AND (
        disconnect_at IS NOT NULL
        OR (jwt_exp < NOW() AND disconnect_at IS NULL)
          );
`;
    const params = [{ value: lt }];
    const result = await this.db.query({ statement: query, params });
    console.log(result.rows);
  }

  async [Symbol.asyncDispose]() {
    await this.db[Symbol.asyncDispose]();
  }

  async prepareStatements(connection: pg_wire.PgConnection) {
    // It should be possible to prepare statements for some common operations here.
    // This has not been implemented yet.
  }
}
