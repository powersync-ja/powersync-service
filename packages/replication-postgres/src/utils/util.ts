import * as pgwire from '@powersync/service-jpgwire';
import { replication } from '@powersync/service-core';

import { ReplicaIdentityResult } from '../types.js';
import { retriedQuery } from './pgwire_utils.js';

export function getReplicaIdColumns(relation: pgwire.PgoutputRelation): replication.ReplicationColumn[] {
  if (relation.replicaIdentity == 'nothing') {
    return [];
  } else {
    return relation.columns.filter((c) => (c.flags & 0b1) != 0).map((c) => ({ name: c.name, typeOid: c.typeOid }));
  }
}

export function getRelId(source: pgwire.PgoutputRelation): number {
  // Source types are wrong here
  const relId = (source as any).relationOid as number;
  if (relId == null || typeof relId != 'number') {
    throw new Error(`No relation id!`);
  }
  return relId;
}

export function getPgOutputRelation(source: pgwire.PgoutputRelation): replication.Relation {
  return {
    name: source.name,
    schema: source.schema,
    relationId: getRelId(source),
    replicaIdentity: source.replicaIdentity,
    replicationColumns: getReplicaIdColumns(source)
  };
}

export async function getPrimaryKeyColumns(
  db: pgwire.PgClient,
  relationId: number,
  mode: 'primary' | 'replident'
): Promise<replication.ReplicationColumn[]> {
  const indexFlag = mode == 'primary' ? `i.indisprimary` : `i.indisreplident`;
  const attrRows = await retriedQuery(db, {
    statement: `SELECT a.attname as name, a.atttypid as typeid, a.attnum as attnum
                                    FROM pg_index i
                                             JOIN pg_attribute a
                                                  ON a.attrelid = i.indrelid
                                                      AND a.attnum = ANY (i.indkey)
                                    WHERE i.indrelid = $1::oid
                                      AND ${indexFlag}
                                      AND a.attnum > 0
                                      ORDER BY a.attnum`,
    params: [{ value: relationId, type: 'int4' }]
  });

  return attrRows.rows.map((row) => {
    return { name: row[0] as string, typeOid: row[1] as number };
  });
}

export async function getAllColumns(db: pgwire.PgClient, relationId: number): Promise<replication.ReplicationColumn[]> {
  const attrRows = await retriedQuery(db, {
    statement: `SELECT a.attname as name, a.atttypid as typeid, a.attnum as attnum
                                    FROM pg_attribute a
                                    WHERE a.attrelid = $1::oid
                                      AND attnum > 0
                                    ORDER BY a.attnum`,
    params: [{ type: 'varchar', value: relationId }]
  });
  return attrRows.rows.map((row) => {
    return { name: row[0] as string, typeOid: row[1] as number };
  });
}

export async function getReplicationIdentityColumns(
  db: pgwire.PgClient,
  relationId: number
): Promise<ReplicaIdentityResult> {
  const rows = await retriedQuery(db, {
    statement: `SELECT CASE relreplident
        WHEN 'd' THEN 'default'
        WHEN 'n' THEN 'nothing'
        WHEN 'f' THEN 'full'
        WHEN 'i' THEN 'index'
     END AS replica_identity
FROM pg_class
WHERE oid = $1::oid LIMIT 1`,
    params: [{ type: 'int8', value: relationId }]
  });
  const idType: string = rows.rows[0]?.[0];
  if (idType == 'nothing' || idType == null) {
    return { replicationIdentity: 'nothing', columns: [] };
  } else if (idType == 'full') {
    return { replicationIdentity: 'full', columns: await getAllColumns(db, relationId) };
  } else if (idType == 'default') {
    return { replicationIdentity: 'default', columns: await getPrimaryKeyColumns(db, relationId, 'primary') };
  } else if (idType == 'index') {
    return { replicationIdentity: 'index', columns: await getPrimaryKeyColumns(db, relationId, 'replident') };
  } else {
    return { replicationIdentity: 'nothing', columns: [] };
  }
}
