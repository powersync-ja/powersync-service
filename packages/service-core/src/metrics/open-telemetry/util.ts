import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-http';
import { PrometheusExporter } from '@opentelemetry/exporter-prometheus';
import { MeterProvider, MetricReader, PeriodicExportingMetricReader } from '@opentelemetry/sdk-metrics';
import { logger } from '@powersync/lib-services-framework';
import { ServiceContext } from '../../system/ServiceContext.js';
import { MetricsFactory } from '../metrics-interfaces.js';
import { OpenTelemetryMetricsFactory } from './OpenTelemetryMetricsFactory.js';

import pkg from '../../../package.json' with { type: 'json' };
import { resourceFromAttributes } from '@opentelemetry/resources';

export function createOpenTelemetryMetricsFactory(context: ServiceContext): MetricsFactory {
  const { configuration, lifeCycleEngine, storageEngine } = context;
  const configuredExporters: MetricReader[] = [];

  if (configuration.telemetry.prometheus_port) {
    const prometheusExporter = new PrometheusExporter({
      port: configuration.telemetry.prometheus_port,
      preventServerStart: true
    });
    configuredExporters.push(prometheusExporter);

    lifeCycleEngine.withLifecycle(prometheusExporter, {
      start: async () => {
        await prometheusExporter.startServer();
        logger.info(`Prometheus metric export enabled on port:${configuration.telemetry.prometheus_port}`);
      }
    });
  }

  if (!configuration.telemetry.disable_telemetry_sharing) {
    const periodicExporter = new PeriodicExportingMetricReader({
      exporter: new OTLPMetricExporter({
        url: configuration.telemetry.internal_service_endpoint
      }),
      exportIntervalMillis: 1000 * 60 * 5 // 5 minutes
    });

    configuredExporters.push(periodicExporter);
  }

  let resolvedInstanceId: (id: string) => void;
  const instanceIdPromise = new Promise<string>((resolve) => {
    resolvedInstanceId = resolve;
  });

  lifeCycleEngine.withLifecycle(null, {
    start: async () => {
      const bucketStorage = storageEngine.activeBucketStorage;
      try {
        const instanceId = await bucketStorage.getPowerSyncInstanceId();
        resolvedInstanceId(instanceId);
      } catch (err) {
        resolvedInstanceId('Unknown');
      }
    }
  });

  const resource = resourceFromAttributes({
    ['service']: 'PowerSync',
    ['service.version']: pkg.version,
    ['instance_id']: instanceIdPromise
  });

  // This triggers OpenTelemetry to resolve the async attributes (instanceIdPromise).
  // This will never reject, and we don't specifically need to wait for it.
  resource.waitForAsyncAttributes?.();

  const meterProvider = new MeterProvider({
    resource,

    readers: configuredExporters
  });

  lifeCycleEngine.withLifecycle(meterProvider, {
    stop: async () => {
      await meterProvider.shutdown();
    }
  });

  const meter = meterProvider.getMeter('powersync');

  return new OpenTelemetryMetricsFactory(meter);
}
