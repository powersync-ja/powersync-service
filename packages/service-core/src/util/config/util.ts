import * as urijs from 'uri-js';

export function getSupabaseJwksUrl(connection: any) {
  if (connection == null) {
    return null;
  } else if (connection.type != 'postgresql') {
    return null;
  }

  let hostname: string | undefined = connection.hostname;
  if (hostname == null && typeof connection.uri == 'string') {
    hostname = urijs.parse(connection.uri).host;
  }
  if (hostname == null) {
    return null;
  }

  const match = /db.(\w+).supabase.co/.exec(hostname);
  if (match == null) {
    return null;
  }
  const projectId = match[1];

  return `https://${projectId}.supabase.co/auth/v1/.well-known/jwks.json`;
}
