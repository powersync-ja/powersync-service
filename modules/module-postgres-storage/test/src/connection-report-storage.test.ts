import { afterAll, beforeAll, describe, expect, it } from 'vitest';
import { POSTGRES_REPORT_STORAGE_FACTORY } from './util.js';
import { event_types } from '@powersync/service-types';
import { register, ReportUserData } from '@powersync/service-core-tests';
import { PostgresReportStorageFactory } from '../../src/storage/PostgresReportStorageFactory.js';
import { DateTimeValue } from '@powersync/service-sync-rules';

const factory = await POSTGRES_REPORT_STORAGE_FACTORY();
const userData = register.REPORT_TEST_USERS;
const dates = register.REPORT_TEST_DATES;

function removeVolatileFields(sdks: event_types.ClientConnection[]): Partial<event_types.ClientConnection>[] {
  return sdks.map((sdk) => {
    const { id, disconnected_at, connected_at, jwt_exp, ...rest } = sdk;
    return {
      ...rest
    };
  });
}

async function loadData(userData: ReportUserData, factory: PostgresReportStorageFactory) {
  await factory.db.sql`
      INSERT INTO
        connection_report_events (
          user_id,
          client_id,
          connected_at,
          sdk,
          user_agent,
          jwt_exp,
          id,
          disconnected_at
        )
      VALUES
        (
          ${{ type: 'varchar', value: userData.user_one.user_id }},
          ${{ type: 'varchar', value: userData.user_one.client_id }},
          ${{ type: 1184, value: userData.user_one.connected_at.toISOString() }},
          ${{ type: 'varchar', value: userData.user_one.sdk }},
          ${{ type: 'varchar', value: userData.user_one.user_agent }},
          ${{ type: 1184, value: userData.user_one.jwt_exp.toISOString() }},
          ${{ type: 'varchar', value: '1' }},
          NULL
        ),
        (
          ${{ type: 'varchar', value: userData.user_two.user_id }},
          ${{ type: 'varchar', value: userData.user_two.client_id }},
          ${{ type: 1184, value: userData.user_two.connected_at.toISOString() }},
          ${{ type: 'varchar', value: userData.user_two.sdk }},
          ${{ type: 'varchar', value: userData.user_two.user_agent }},
          ${{ type: 1184, value: userData.user_two.jwt_exp.toISOString() }},
          ${{ type: 'varchar', value: '2' }},
          NULL
        ),
        (
          ${{ type: 'varchar', value: userData.user_four.user_id }},
          ${{ type: 'varchar', value: userData.user_four.client_id }},
          ${{ type: 1184, value: userData.user_four.connected_at.toISOString() }},
          ${{ type: 'varchar', value: userData.user_four.sdk }},
          ${{ type: 'varchar', value: userData.user_four.user_agent }},
          ${{ type: 1184, value: userData.user_four.jwt_exp.toISOString() }},
          ${{ type: 'varchar', value: '4' }},
          NULL
        ),
        (
          ${{ type: 'varchar', value: userData.user_old.user_id }},
          ${{ type: 'varchar', value: userData.user_old.client_id }},
          ${{ type: 1184, value: userData.user_old.connected_at.toISOString() }},
          ${{ type: 'varchar', value: userData.user_old.sdk }},
          ${{ type: 'varchar', value: userData.user_old.user_agent }},
          ${{ type: 1184, value: userData.user_old.jwt_exp.toISOString() }},
          ${{ type: 'varchar', value: '5' }},
          NULL
        ),
        (
          ${{ type: 'varchar', value: userData.user_three.user_id }},
          ${{ type: 'varchar', value: userData.user_three.client_id }},
          ${{ type: 1184, value: userData.user_three.connected_at.toISOString() }},
          ${{ type: 'varchar', value: userData.user_three.sdk }},
          ${{ type: 'varchar', value: userData.user_three.user_agent }},
          NULL,
          ${{ type: 'varchar', value: '3' }},
          ${{ type: 1184, value: userData.user_three.disconnected_at.toISOString() }}
        ),
        (
          ${{ type: 'varchar', value: userData.user_week.user_id }},
          ${{ type: 'varchar', value: userData.user_week.client_id }},
          ${{ type: 1184, value: userData.user_week.connected_at.toISOString() }},
          ${{ type: 'varchar', value: userData.user_week.sdk }},
          ${{ type: 'varchar', value: userData.user_week.user_agent }},
          NULL,
          ${{ type: 'varchar', value: 'week' }},
          ${{ type: 1184, value: userData.user_week.disconnected_at.toISOString() }}
        ),
        (
          ${{ type: 'varchar', value: userData.user_month.user_id }},
          ${{ type: 'varchar', value: userData.user_month.client_id }},
          ${{ type: 1184, value: userData.user_month.connected_at.toISOString() }},
          ${{ type: 'varchar', value: userData.user_month.sdk }},
          ${{ type: 'varchar', value: userData.user_month.user_agent }},
          NULL,
          ${{ type: 'varchar', value: 'month' }},
          ${{ type: 1184, value: userData.user_month.disconnected_at.toISOString() }}
        ),
        (
          ${{ type: 'varchar', value: userData.user_expired.user_id }},
          ${{ type: 'varchar', value: userData.user_expired.client_id }},
          ${{ type: 1184, value: userData.user_expired.connected_at.toISOString() }},
          ${{ type: 'varchar', value: userData.user_expired.sdk }},
          ${{ type: 'varchar', value: userData.user_expired.user_agent }},
          ${{ type: 1184, value: userData.user_expired.jwt_exp.toISOString() }},
          ${{ type: 'varchar', value: 'expired' }},
          NULL
        )
    `.execute();
}

async function deleteData(factory: PostgresReportStorageFactory) {
  await factory.db.sql`TRUNCATE TABLE connection_report_events`.execute();
}

beforeAll(async () => {
  await loadData(userData, factory);
});
afterAll(async () => {
  await deleteData(factory);
});

describe('Report storage tests', async () => {
  await register.registerReportTests(factory);
});

describe('Connection report storage', async () => {
  it('Should update a connection event if its within a day', async () => {
    const newConnectAt = new Date(
      dates.now.getFullYear(),
      dates.now.getMonth(),
      dates.now.getDate(),
      dates.now.getHours(),
      dates.now.getMinutes() + 20
    );
    const jwtExp = new Date(newConnectAt.getFullYear(), newConnectAt.getMonth(), newConnectAt.getDate() + 1);
    await factory.reportClientConnection({
      sdk: userData.user_one.sdk,
      connected_at: newConnectAt,
      jwt_exp: jwtExp,
      client_id: userData.user_one.client_id,
      user_id: userData.user_one.user_id,
      user_agent: userData.user_one.user_agent
    });

    const sdk = await factory.db
      .sql`SELECT * FROM connection_report_events WHERE user_id = ${{ type: 'varchar', value: userData.user_one.user_id }} AND client_id = ${{ type: 'varchar', value: userData.user_one.client_id }}`.rows<event_types.ClientConnection>();
    expect(sdk).toHaveLength(1);
    expect(new Date((sdk[0].connected_at as unknown as DateTimeValue).iso8601Representation).toISOString()).toEqual(
      newConnectAt.toISOString()
    );
    expect(new Date((sdk[0].jwt_exp! as unknown as DateTimeValue).iso8601Representation).toISOString()).toEqual(
      jwtExp.toISOString()
    );
    expect(sdk[0].disconnected_at).toBeNull();
    const cleaned = removeVolatileFields(sdk);
    expect(cleaned).toMatchSnapshot();
  });

  it('Should update a connection event and make it disconnected', async () => {
    const disconnectAt = new Date(
      dates.now.getFullYear(),
      dates.now.getMonth(),
      dates.now.getDate(),
      dates.now.getHours(),
      dates.now.getMinutes() + 20
    );
    const jwtExp = new Date(disconnectAt.getFullYear(), disconnectAt.getMonth(), disconnectAt.getDate() + 1);

    await factory.reportClientDisconnection({
      disconnected_at: disconnectAt,
      jwt_exp: jwtExp,
      client_id: userData.user_three.client_id,
      user_id: userData.user_three.user_id,
      user_agent: userData.user_three.user_agent,
      connected_at: dates.yesterday
    });

    const sdk = await factory.db
      .sql`SELECT * FROM connection_report_events WHERE user_id = ${{ type: 'varchar', value: userData.user_three.user_id }}`.rows<event_types.ClientConnection>();
    expect(sdk).toHaveLength(1);
    console.log(sdk[0]);
    expect(new Date((sdk[0].disconnected_at! as unknown as DateTimeValue).iso8601Representation).toISOString()).toEqual(
      disconnectAt.toISOString()
    );
    const cleaned = removeVolatileFields(sdk);
    expect(cleaned).toMatchSnapshot();
  });

  it('Should create a connection event if its after a day', async () => {
    const newConnectAt = new Date(
      dates.now.getFullYear(),
      dates.now.getMonth(),
      dates.now.getDate() + 1,
      dates.now.getHours()
    );
    const jwtExp = new Date(newConnectAt.getFullYear(), newConnectAt.getMonth(), newConnectAt.getDate() + 1);

    await factory.reportClientConnection({
      sdk: userData.user_week.sdk,
      connected_at: newConnectAt,
      jwt_exp: jwtExp,
      client_id: userData.user_week.client_id,
      user_id: userData.user_week.user_id,
      user_agent: userData.user_week.user_agent
    });

    const sdk = await factory.db
      .sql`SELECT * FROM connection_report_events WHERE user_id = ${{ type: 'varchar', value: userData.user_week.user_id }}`.rows<event_types.ClientConnection>();
    expect(sdk).toHaveLength(2);
    const cleaned = removeVolatileFields(sdk);
    expect(cleaned).toMatchSnapshot();
  });

  it('Should delete rows older than specified range', async () => {
    await deleteData(factory);
    await loadData(userData, factory);
    await factory.deleteOldConnectionData({
      date: dates.weekAgo
    });
    const sdk = await factory.getClientConnectionReports({
      start: dates.monthAgo,
      end: dates.now
    });
    expect(sdk).toMatchSnapshot();
  });
});
