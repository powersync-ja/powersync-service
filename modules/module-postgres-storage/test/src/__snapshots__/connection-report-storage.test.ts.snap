// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Connection report storage > Should create a connection event if its after a day 1`] = `
[
  {
    "client_id": "client_week",
    "sdk": "powersync-js/1.24.5",
    "user_agent": "powersync-js/1.21.0 powersync-web Firefox/141 linux",
    "user_id": "user_week",
  },
  {
    "client_id": "client_week",
    "sdk": "powersync-js/1.24.5",
    "user_agent": "powersync-js/1.21.0 powersync-web Firefox/141 linux",
    "user_id": "user_week",
  },
]
`;

exports[`Connection report storage > Should create a sdk event if its after a day 1`] = `
[
  {
    "client_id": "client_week",
    "sdk": "powersync-js/1.24.5",
    "user_agent": "powersync-js/1.21.0 powersync-web Firefox/141 linux",
    "user_id": "user_week",
  },
  {
    "client_id": "client_week",
    "sdk": "powersync-js/1.24.5",
    "user_agent": "powersync-js/1.21.0 powersync-web Firefox/141 linux",
    "user_id": "user_week",
  },
]
`;

exports[`Connection report storage > Should delete rows older than specified range 1`] = `
{
  "sdks": [
    {
      "clients": 1,
      "sdk": "powersync-dart/1.6.4",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.1",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.2",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.4",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.24.5",
      "users": 1,
    },
  ],
  "users": 5,
}
`;

exports[`Connection report storage > Should show connected users with start range 1`] = `
{
  "sdks": [
    {
      "clients": 1,
      "sdk": "powersync-dart/1.6.4",
      "users": 1,
    },
  ],
  "users": 1,
}
`;

exports[`Connection report storage > Should show connected users with start range and end range 1`] = `
{
  "sdks": [
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.1",
      "users": 1,
    },
  ],
  "users": 1,
}
`;

exports[`Connection report storage > Should show connection report data for user over the past day 1`] = `
{
  "sdks": [
    {
      "clients": 1,
      "sdk": "powersync-dart/1.6.4",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.1",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.4",
      "users": 1,
    },
  ],
  "users": 3,
}
`;

exports[`Connection report storage > Should show connection report data for user over the past month 1`] = `
{
  "sdks": [
    {
      "clients": 1,
      "sdk": "powersync-dart/1.6.4",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.1",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.2",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.4",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.23.6",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.23.7",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.24.5",
      "users": 1,
    },
  ],
  "users": 7,
}
`;

exports[`Connection report storage > Should show connection report data for user over the past week 1`] = `
{
  "sdks": [
    {
      "clients": 1,
      "sdk": "powersync-dart/1.6.4",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.1",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.2",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.21.4",
      "users": 1,
    },
    {
      "clients": 1,
      "sdk": "powersync-js/1.24.5",
      "users": 1,
    },
  ],
  "users": 5,
}
`;

exports[`Connection report storage > Should update a connected sdk event and make it disconnected 1`] = `
[
  {
    "client_id": "client_three",
    "sdk": "powersync-js/1.21.2",
    "user_agent": "powersync-js/1.21.0 powersync-web Firefox/141 linux",
    "user_id": "user_three",
  },
]
`;

exports[`Connection report storage > Should update a connection event and make it disconnected 1`] = `
[
  {
    "client_id": "client_three",
    "sdk": "powersync-js/1.21.2",
    "user_agent": "powersync-js/1.21.0 powersync-web Firefox/141 linux",
    "user_id": "user_three",
  },
]
`;

exports[`Connection report storage > Should update a connection event if its within a day 1`] = `
[
  {
    "client_id": "client_one",
    "sdk": "powersync-dart/1.6.4",
    "user_agent": "powersync-dart/1.6.4 Dart (flutter-web) Chrome/128 android",
    "user_id": "user_one",
  },
]
`;

exports[`Connection report storage > Should update a sdk event if its within a day 1`] = `
[
  {
    "client_id": "client_one",
    "sdk": "powersync-dart/1.6.4",
    "user_agent": "powersync-dart/1.6.4 Dart (flutter-web) Chrome/128 android",
    "user_id": "user_one",
  },
]
`;
